### PR TITLE
miri: avoid making a full copy of all new allocations

### LINKED ISSUE #3637: Big `Vec::try_reserve` OOMs Miri
On both GitHub CI default linux runner (should have 16GB RAM), and locally (also linux, 64GB RAM), the following test will run out of memory in Miri, but panic in normal `cargo test`:

```rust
#[test]
fn try_allocate_a_lot() {
    let mut v = Vec::<u8>::new();
    v.try_reserve(128 * 1024 * 1024 * 1024).unwrap();
    let _ = v;
}
```

- `cargo test` panics: ``called `Result::unwrap()` on an `Err` value: TryReserveError { kind: AllocError { layout: Layout { size: 137438953472, align: 1 (1 << 0) }, non_exhaustive: () } }``
- `cargo miri test` will try allocating and get killed for OOM, CTRL+C does not stop it

Apologies if this has been reported before, I could not find another issue for this.

### PR DESCRIPTION
**Pull request by RalfJung**
Hopefully fixes https://github.com/rust-lang/miri/issues/3637

r? @saethlin

### CODE REVIEW

File: compiler/rustc_const_eval/src/interpret/machine.rs [hunk 1]
```diff
@@ -329,27 +329,41 @@ pub trait Machine<'tcx>: Sized {
...
-        alloc: Cow<'b, Allocation>,
-        kind: Option<MemoryKind<Self::MemoryKind>>,
-    ) -> InterpResult<'tcx, Cow<'b, Allocation<Self::Provenance, Self::AllocExtra, Self::Bytes>>>;
+        alloc: &'b Allocation,
+    ) -> InterpResult<'tcx, Cow<'b, Allocation<Self::Provenance, Self::AllocExtra, Self::Bytes>>>
+    {
+        // The default implementation does a copy; CTFE machines have a more efficient implementation
+        // based on their particular choice for `Provenance`, `AllocExtra`, and `Bytes`.
+        let kind = Self::GLOBAL_KIND
+            .expect("if GLOBAL_KIND is None, adjust_global_allocation must be overwritten");
```
Comment (saethlin) [inline compiler/rustc_const_eval/src/interpret/machine.rs:349]: :raised_eyebrow: So the implementation that Miri needs is the default, and the `const fn` interpreter overrides with something simpler? Don't we usually do things the other way around?

Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/machine.rs:349]: Yeah, but the other way around does not work here and the const-eval interpreter impl is not well-typed in general, in relies on specific choices for the associated type. And meanwhile it seems nice that a default impl is possible.

Comment (saethlin) [reply compiler/rustc_const_eval/src/interpret/machine.rs:349]: I see. It's always ironic when we run into loose typing implementing Rust.

Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/machine.rs:349]: To me this is actually a sign of a pretty powerful type system -- in the const-eval implementation we of course know more about the concrete choice of types than in the general trait definition, and that lets us safely convert between types that are in general inequal, but are known to be equal in this concrete instance.

**Pull request merged**